_posts/2020-08-01-data-science-meets-devops.md

```diff
@@ -0,0 +1,769 @@
+   </td>
+  </tr>
+</table>
+
+
+The first controller, the Trainer, (left hand side of Figure 2) is a controller which checks if our model needs to be retrained. At the moment, this controller retrains the model after a set time period. When the model becomes stale, the controller  executes a [notebook ](https://github.com/kubeflow/code-intelligence/blob/master/Label_Microservice/notebooks/automl.ipynb) programmatically using [papermill.](https://github.com/nteract/papermill)  This notebook fetches GitHub Issues data [from BigQuery](https://medium.com/google-cloud/analyzing-github-issues-and-comments-with-bigquery-c41410d3308) and launches an [AutoML](https://cloud.google.com/automl) job to train a model. 
+
+AutoML allowed us to focus on infrastructure and engineering rather than spending too much time building and tuning models.  Furthermore, AutoML provides a competitive baseline for us to improve upon.  We may revisit these models in the future.
+
+The second component, the Deployer, (right hand side of Figure 2) determines which model should be live and if needed, automatically opens a pull request to update our model. Once the pull request is merged [Anthos Config Mesh](https://cloud.google.com/anthos/config-management) automatically rolls it out to production.
+
+
+## Further Details
+
+
+### Background
+
+Before diving into the details of our solution we want to provide some context around:
+
+
+
+1. What reconcilers are and why we use them
+2. What GitOps is and why we adopted this pattern
+
+
+### Building Resilient Systems With Reconcilers
+
+A reconciler is a control pattern that has proven to be immensely useful for building resilient systems. The reconcile pattern is [at the heart of how Kubernetes works](https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html). Figure 3 illustrates how a reconciler works. A reconciler works by first observing the state of the world; e.g. what model is currently deployed. The reconciler then compares this against the desired state of the world and computes the diff; e.g the model with label “version=20200724” should be deployed, but the model currently deployed has label “version=20200700”. The reconciler then takes the action necessary to drive the world to the desired state; e.g. open a pull request to change the deployed model.
+
+
+<table>
+  <tr>
+   <td>
+<img src="/images/2020-08-01-data-science-meets-devops/fig3.reconciler.png" width="" alt="alt_text" title="image_tooltip">
+
+   </td>
+  </tr>
+  <tr>
+   <td>Figure 3. Illustration of the reconciler pattern as applied by our deployer.
+   </td>
+  </tr>
+</table>
+
+
+Reconcilers have proven immensely useful for building resilient systems because a well implemented reconciler provides a high degree of confidence that no matter how a system is perturbed it will eventually return to the desired state.
```

Comment:
Might be worth pointing to some control theory or other reconciler implementations (i.e. kubernetes itself)